src/Components/test/E2ETest/ServerExecutionTests/CircuitGracefulTerminationTests.cs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using BasicTestApp;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests;
+using Microsoft.AspNetCore.E2ETesting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Testing;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Chrome;
+using OpenQA.Selenium.Remote;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETests.ServerExecutionTests
+{
+    public class CircuitGracefulTerminationTests : BasicTestAppTestBase, IDisposable
+    {
+        public CircuitGracefulTerminationTests(
+            BrowserFixture browserFixture,
+            ToggleExecutionModeServerFixture<Program> serverFixture,
+            ITestOutputHelper output)
+            : base(browserFixture, serverFixture.WithServerExecution(), output)
+        {
+        }
+
+        public TaskCompletionSource<object> GracefulDisconnectCompletionSource { get; private set; }
+        public TestSink Sink { get; private set; }
+        public List<(Extensions.Logging.LogLevel level, string eventIdName)> Messages { get; private set; }
+
+        protected override void InitializeAsyncCore()
+        {
+            // On WebAssembly, page reloads are expensive so skip if possible
+            Navigate(ServerPathBase, _serverFixture.ExecutionMode == ExecutionMode.Client);
+            MountTestComponent<CounterComponent>();
+            Browser.Equal("Current count: 0", () => Browser.FindElement(By.TagName("p")).Text);
+        }
+
+        [Fact]
+        public async Task ReloadingThePage_GracefullyDisconnects_TheCurrentCircuit()
+        {
+            // Arrange
+            GracefulDisconnectCompletionSource = new TaskCompletionSource<object>(TaskContinuationOptions.RunContinuationsAsynchronously);
+            Sink = _serverFixture.Host.Services.GetRequiredService<TestSink>();
+            Messages = new List<(Extensions.Logging.LogLevel level, string eventIdName)>();
+            Sink.MessageLogged += wc => Log(wc);
+
+            // Act
+            _ = ((IJavaScriptExecutor)Browser).ExecuteScript("location.reload()");
+            await Task.WhenAny(Task.Delay(10000), GracefulDisconnectCompletionSource.Task);
+
+            // Assert
+            Assert.Contains((Extensions.Logging.LogLevel.Debug, "CircuitTerminatedGracefully"), Messages);
```

Comment:
I'm not sure we need to assert about the contents of `Messages`. Doesn't the `await` on L56 already say the same thing, i.e., that we saw a `CircuitTerminatedGracefully` log entry?

Reply:
Yeah, but this results in a clear exception when it fails.

Reply:
Oh I see, hence the `WhenAny`. OK cool!

Reply:
It's for clarity, the task above is either going to wait for 10 seconds or until it sees the message for the first time. When it doesn't see the message there needs to be an assertion. 

This way, if after 10 seconds we don't see the message, we simply throw on the assert and the failure is clear.